lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -272,13 +294,68 @@ static LogicalResult verifyCellOp(CellOp cell) {
 //===----------------------------------------------------------------------===//
 static LogicalResult verifyAssignOp(AssignOp assign) {
   auto parent = assign->getParentOp();
-  if (!(isa<GroupOp>(parent) || isa<WiresOp>(parent)))
+  if (!isa<GroupOp, WiresOp>(parent))
     return assign.emitOpError(
         "should only be contained in 'calyx.wires' or 'calyx.group'");
 
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// ControlOp
+//===----------------------------------------------------------------------===//
+
+/// A helper function to verify that each operation in
+/// the body of a control-like operation is valid.
+static LogicalResult verifyControlLikeOpBody(Operation *op) {
+  assert(op->getNumRegions() != 0 && "The operation should have a region.");
+  auto &region = op->getRegion(0);
+  assert(region.hasOneBlock() && "The region should have one block.");
+
+  bool isNotControlOp = !isa<ControlOp>(op);
+  for (auto &&bodyOp : region.front()) {
+    if (isa<SeqOp>(bodyOp))
```

Comment:
So this is to check for ops that are allowed in the control section? Seems to me this could be better expressed by a trait, like `ControlSectionOp` or something, and the check here would just ensure that all ops had that trait.

Reply:
Also, this doesn't recurse into nested regions, and it appears SeqOp  can be nested. In that case, I might use `op->walk(...)`, which should recurse through nested regions.